### PR TITLE
fix: remove `const` from the argument `chip::Span<chip::DataVersion>` of `emberAfSetDynamicEndpoint()` because there is internal changes

### PR DESCRIPTION
**Pull request by hideakitai**
In the arguments of `emberAfSetDynamicEndpoint()`,

https://github.com/project-chip/connectedhomeip/blob/4855f580034348e98a4c4259c6d50f784a432c36/src/app/util/attribute-storage.cpp#L242-L244

`chip::Span<chip::DataVersion>` is marked as `const`, but internal changes exist here.

https://github.com/project-chip/connectedhomeip/blob/4855f580034348e98a4c4259c6d50f784a432c36/src/app/util/attribute-storage.cpp#L286

So the argument `const chip::Span<chip::DataVersion> & dataVersionStorage` should not be `const`.

### CODE REVIEW

File: examples/bridge-app/asr/subdevice/SubDeviceManager.cpp [hunk 1]
```diff
@@ -48,7 +48,7 @@ static EndpointId gFirstDynamicEndpointId;
 static SubDevice * gSubDevices[CHIP_DEVICE_CONFIG_DYNAMIC_ENDPOINT_COUNT]; // number of dynamic endpoints count
 
 int AddDeviceEndpoint(SubDevice * dev, EmberAfEndpointType * ep, const Span<const EmberAfDeviceType> & deviceTypeList,
-                      const Span<DataVersion> & dataVersionStorage, chip::EndpointId parentEndpointId)
+                      Span<DataVersion> && dataVersionStorage, chip::EndpointId parentEndpointId)
```
Comment (bzbarsky-apple) [inline examples/bridge-app/asr/subdevice/SubDeviceManager.cpp:51]: Why is this an rvalue?  Doesn't that allow caller to pass a temporary?  `dataVersionStorage` can't be a temporary...

File: src/app/util/attribute-storage.h [hunk 2]
```diff
@@ -249,7 +249,7 @@ CHIP_ERROR SetTagList(chip::EndpointId endpoint,
 // An optional parent endpoint id should be passed for child endpoints of composed device.
 //
 EmberAfStatus emberAfSetDynamicEndpoint(uint16_t index, chip::EndpointId id, const EmberAfEndpointType * ep,
-                                        const chip::Span<chip::DataVersion> & dataVersionStorage,
+                                        chip::Span<chip::DataVersion> & dataVersionStorage,
```
Comment (bzbarsky-apple) [inline src/app/util/attribute-storage.h:252]: I'm not sure this is right.  The Span itself is const: the callee does not mutate its length or what it points to.

The data in the buffer is modified, but that's fine; this is not a `Span<const DataVersion>`; it's a `const Span<DataVersion>`, no?

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]

**Comment by bzbarsky-apple** [comment]
An in particular, this line that's quoted:
```
 if (Crypto::DRBG_get_bytes(reinterpret_cast<uint8_t *>(dataVersionStorage.data()), dataSize) != CHIP_NO_ERROR) 
```
compiles precisely because `dataVersionStorage.data()` is _not_ a pointer to const data.  If it were, the compiler would fail the cast, because `reinterpret_cast` cannot cast away constness (unlike a C-style cast).

**Comment by hideakitai** [comment]
@bzbarsky-apple  Thank you for your comment. You are correct in what you say.

The problem in my environment is when the `DataVersion` of `chip::Span<DataVersion>` is `constexpr`. In this case, the compile passes fine, but the write to the `DataVersion` which is `constexpr` causes panic at runtime.

After your comment, I tried to think of a way to prevent this from happening at compile time, but I could have come up with a better idea. Do you have any good ideas?

**Comment by bzbarsky-apple** [comment]
@hideakitai Hmm.  If that DataVersion is constexpr, how are we ending up with a non-const pointer to it at all?  What does the code on the caller side that creates the Span look like in this case?

**Comment by bzbarsky-apple** [comment]
@hideakitai Ah, so this line:
```
            (chip::DataVersion*)this->DATA_VERSIONS.data(), 
```
is a C-style cast that casts away const.  And once you are casting away const, and passing const things to something that expects non-const, all bets are off.

You'd get the same problems if you C-style cast a random integer to a pointer or whatnot.

You really shouldn't use C-style casts in anything resembling modern C++.  The fact that one was needed here to get this code to compile should have been a giant red flag.

**Comment by hideakitai** [comment]
@bzbarsky-apple It was right in front of my eyes, but I completely missed it...that's exactly right. I had completely forgotten about using a C-style cast because I had not passed this part of the compilation for a long time. Thank you very much.

---

By the way, `chip::Span<T>` has the following constructor that takes `std::array<T, N>` (not `const`) as an argument.

```c++
    template <class U, size_t N, typename = std::enable_if_t<std::is_same<std::remove_const_t<T>, std::remove_const_t<U>>::value>>
    constexpr Span(std::array<U, N> & arr) : mDataBuf(arr.data()), mDataLen(N)
    {}
```

However, if I pass a const std::array to it (`this->DEVICE_TYPES` is `constexpr std::array<const EmberAfDeviceType, 1>`), 

```c++
        .device_types = chip::Span<const EmberAfDeviceType>(this->DEVICE_TYPES),
```

I get the following template deduction/substitution error. Is this because `const std::array<U, N>` is not supported by the constructor of `chip::Span<T>` ?

```
connectedhomeip/src/lib/support/Span.h:50:15: note: candidate: 'template<class U, unsigned int N, class> constexpr chip::Span<T>::Span(std::array<U, N>&)'
     constexpr Span(std::array<U, N> & arr) : mDataBuf(arr.data()), mDataLen(N)
               ^~~~
connectedhomeip/src/lib/support/Span.h:50:15: note:   template argument deduction/substitution failed:
matter_core_endpoint.cpp:328:79: note:   types 'std::array<_Tp, _Nm>' and 'const std::array<const EmberAfDeviceType, 1>' have incompatible cv-qualifiers
         .device_types = chip::Span<const EmberAfDeviceType>(this->DEVICE_TYPES),
```

**Comment by hideakitai** [comment]
Ah I got it. I'm using v1.0.0.2, but the constructors are fixed to accept `const std::array<T, N>` in mainstream.

https://github.com/project-chip/connectedhomeip/blob/caea3ec81b8c1f1e690323a00426301dd9c9a05a/src/lib/support/Span.h#L80-L91

**Comment by hideakitai** [comment]
Anyway, sorry for making you spend your time debugging my code. It was really helpful, thank you very much!

**Comment by bzbarsky-apple** [comment]
You're welcome!

**Pull request closed without merging**